### PR TITLE
WIP: Kruskal MST Algorithm updated

### PR DESCRIPTION
**Pull request by athulappadan**
This is a follow up of an earlier PR. (#456)

### CODE REVIEW

File: src/spanningtrees/kruskal.jl [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+end
+
+"""Performs [Kruskal's algorithm](https://en.wikipedia.org/wiki/Kruskal%27s_algorithm)
+on a connected, non-directional graph `g`, having adjacency matrix `distmx`,
+and computes minimum spanning tree. Returns a `Vector{KruskalHeapEntry}`,
+that contains the containing edges and its weights.
+"""
+function kruskal_mst{T<:Real}(
+    g::SimpleGraph,
+    distmx::AbstractArray{T, 2}
```
Comment (CarloLucibello) [inline src/spanningtrees/kruskal.jl]: `distmx::AbstractArray{T, 2} = DefaultDistance()`

so that we can call kruskal_mst(g)  and assume weight 1 on each edge. Take a look to src/distance.jl for examples

File: src/spanningtrees/kruskal.jl [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    edge_list = Vector{KruskalHeapEntry{T}}()
+    mst = Vector{KruskalHeapEntry{T}}()
+    connected_nodes = Vector{Int}(1:nv(g))
+
+    sizehint!(edge_list, ne(g))
+    sizehint!(mst, ne(g))
+
+    for e in edges(g)
+        heappush!(edge_list, KruskalHeapEntry{T}(e, distmx[src(e), dst(e)]))
```
Comment (CarloLucibello) [inline src/spanningtrees/kruskal.jl]: just a curiosity: where is defined `heappush!`? i cannot find it

File: src/spanningtrees/kruskal.jl [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        v = src(e.edge)
+        w = dst(e.edge)
+
+        if connected_nodes[v] != connected_nodes[w]
+            quick_find!(connected_nodes, v, w)
+            push!(mst, e)
+        end
+    end
+
+    return mst
```
Comment (CarloLucibello) [inline src/spanningtrees/kruskal.jl]: I'd like the returned object to be a vector containing all the edges in the  mst  instead of this

File: src/spanningtrees/kruskal.jl (hunk 2 above)
Comment (athulappadan) [inline src/spanningtrees/kruskal.jl]: heappush! is a stdlib function, right..?

File: src/spanningtrees/kruskal.jl (hunk 1 above)
Comment (athulappadan) [inline src/spanningtrees/kruskal.jl]: Okay..

File: src/spanningtrees/kruskal.jl (hunk 2 above)
Comment (CarloLucibello) [inline src/spanningtrees/kruskal.jl]: this is weird, in my REPL 
```
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.6 (2016-06-19 17:16 UTC)
 _/ |\__'_|_|_|\__'_|  |  
|__/                   |  x86_64-pc-linux-gnu

julia> heappush!(rand(10))
ERROR: UndefVarError: heappush! not defined
```

Comment (athulappadan) [inline src/spanningtrees/kruskal.jl]: http://docs.julialang.org/en/release-0.5/stdlib/collections/

File: src/spanningtrees/kruskal.jl (hunk 3 above)
Comment (athulappadan) [inline src/spanningtrees/kruskal.jl]: I thought it will be better to include the corresponding weights as well.

Comment (CarloLucibello) [inline src/spanningtrees/kruskal.jl]: if you change the return type to a vector of edge after #464 and this PR : 
```julia
tree, vmap = subgraph(g, kruskal_mst(g))
```
Is there any reason for reaturning the heap?

File: src/spanningtrees/kruskal.jl (hunk 2 above)
Comment (CarloLucibello) [inline src/spanningtrees/kruskal.jl]: ah ok! thanks

File: src/spanningtrees/kruskal.jl (hunk 3 above)
Comment (sbromberger) [inline src/spanningtrees/kruskal.jl]: Be careful. At this point we either hold off on this PR until we've hashed out #464 or we merge without it. The issue is that we don't want to create cascading PR dependencies on a PR (#464) that still needs some serious discussion.

Comment (CarloLucibello) [inline src/spanningtrees/kruskal.jl]: yep
```julia
tree, vmap = subgraph(g, kruskal_mst(g))
```
is not part of this pr, it is just something that could be eventually done user-side if #464 gets merged

Comment (athulappadan) [inline src/spanningtrees/kruskal.jl]: Is Vector{LightGraphs.Edge} okay..?

Comment (sbromberger) [inline src/spanningtrees/kruskal.jl]: You won't need `LightGraphs.Edge` - just `Edge` should suffice.

**Comment by athulappadan** [comment]
All right. Can anyone please explain the steps to update this PR without creating conflicts..?

**Comment by sbromberger** [comment]
Well, at this point there are no conflicts to merge. :)

**Comment by athulappadan** [comment]
Seems like I didn't change the test cases..

**Comment by sbromberger** [comment]
I think the problem is that you're referencing `e.edge` when `e` itself is an edge.

**Comment by athulappadan** [comment]
e is not an edge. And I will rename the variable to avoid confusion..

**Comment by sbromberger** [comment]
> e is not an edge.

Ah, I see it now. It was the `push!()` that got me.

> And I will rename the variable to avoid confusion..

That'd be best. Thanks :)

**Comment by sbromberger** [comment]
@CarloLucibello does this resolve your issue above? If so, are there any reasons we shouldn't merge?

**Comment by CarloLucibello** [comment]
ok, two more things and we are done, thanks for your patience @athulappadan :)

- can you rename the function from kruskal_mst to minimum_spanning_tree? better to go with the latter, since this is the only function we have for finding the minimum spanning tree

- we can add some more tests with little efforts just adapting the ones in #277, could you do that?

beside that, docs need some polishing, put we could do that later

**Comment by sbromberger** [comment]
@athulappadan if you can get this in by, say, 9pm PST, it'll be in 0.7.1. :)

**Comment by athulappadan** [comment]
I will add the tests, and shouldn't we mention `kruskal` in the function name? I was thinking of doing Prim's algorithm after this one..

**Comment by sbromberger** [comment]
Yeah - if there are multiple implementations, we should use the name. We could always alias `minimum_spanning_tree` to one of them.

**Comment by athulappadan** [comment]
@sbromberger I had already slept by that time.. ;)

**Comment by sbromberger** [comment]
You've still got an hour :)

Actually, keep at it. I'm going to hold off tagging a new version since we've got some outstanding bugs to fix.

File: test/spanningtrees/kruskal.jl [hunk 4]
```diff
@@ -40,3 +40,68 @@ push!(vec_mst, Edge(3, 4))
 push!(vec_mst, Edge(2, 3))
 
 @test mst == vec_mst
+
+#second test
+g2 = Graph(8)
```
Comment (sbromberger) [inline test/spanningtrees/kruskal.jl]: Don't override g2 here (it's defined in runtests.jl; we're debating how to move forward with it).

Comment (athulappadan) [inline test/spanningtrees/kruskal.jl]: okay..

Comment (athulappadan) [inline test/spanningtrees/kruskal.jl]: Should I rename it? And are there any other  changes to be made?

Comment (CarloLucibello) [inline test/spanningtrees/kruskal.jl]: we are good, I'll add `minimum_spanning_tree` and some doc fixes in a later PR. Thanks!

Comment (sbromberger) [inline test/spanningtrees/kruskal.jl]: or just reuse `g` if you don't need it anymore.

Comment (sbromberger) [inline test/spanningtrees/kruskal.jl]: Looks great; thanks for sticking with it. I'll merge it tonight assuming it passes tests. :)

**Comment by sbromberger** [comment]
All good! Merging now. Thanks again.

**Comment by athulappadan** [comment]
@sbromberger @CarloLucibello 
Thanks! See you with the next implementation.. :)

**Comment by sbromberger** [comment]
@athulappadan looking forward to it!!!
